CRM/Activity/Form/Task/PDFLetterCommon.php

```diff
@@ -0,0 +1,116 @@
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+use Civi\Token\TokenProcessor;
+
+/**
+ * This class provides the common functionality for creating PDF letter for
+ * activities.
+ *
+ */
+class CRM_Activity_Form_Task_PDFLetterCommon extends CRM_Contact_Form_Task_PDFLetterCommon {
+
+  public static function postProcess(&$form) {
+    self::postProcessActivities($form, $form->_activityHolderIds);
+  }
+
+  /**
+   * Process the form after the input has been submitted and validated.
+   * This uses the new token processor
+   *
+   * @param CRM_Core_Form $form
+   * @param $activityIds
+   *
+   * @return void
+   */
+  public static function postProcessActivities(&$form, $activityIds) {
+    $formValues = $form->controller->exportValues($form->getName());
+    list($formValues, $categories, $html_message, $messageToken, $returnProperties)
+      = self::processMessageTemplate($formValues);
+
+    $tp = new TokenProcessor(\Civi::dispatcher(), array(
+      'controller' => __CLASS__,
+      'smarty' => FALSE,
+    ));
+    $tp->addMessage('body_html', $html_message, 'text/html');
+
+    $activities = civicrm_api3('Activity', 'get', array(
+      'id' => array('IN' => $activityIds),
+    ));
+
+    foreach ($activityIds as $activityId) {
+      $activity = CRM_Utils_Array::value($activityId, $activities['values']);
+      if ($activity) {
+        // NB - must have 'contactId' context otherwise evaluate() blows up
+        $tp->addRow()
+          ->context('activityId', $activityId)
+          ->context('actionSearchResult', (object) $activity)
+          ->context('contactId', $activity['source_contact_id']);
+      }
+    }
+    $tp->evaluate();
+    foreach ($tp->getRows() as $row) {
+      $html[] = $row->render('body_html');
+    }
+
+    // If we were to record a PDF activity, who would we record it against??
+    // probably Target
+    // self::createActivities($form, $html_message, $contactIDs, $formValues['subject'], CRM_Utils_Array::value('campaign_id', $formValues));
+
+    if (!empty($formValues['is_unit_test'])) {
+      return $html;
+    }
+
+    if (!empty($html)) {
```

Comment:
I feel like this section could be shared with another class rather than duplication?  - perhaps from getRows down could be renderHtmlFromRows or similar?